rfc/20250728-opentofu-providers/01-provider-protocol.md

```diff
@@ -0,0 +1,164 @@
+# Provider Protocol
+
+## Summary
+
+This document defines the stdio-based communication protocol for OpenTofu providers, using MessagePack-RPC for efficient, language-agnostic communication. The protocol enables the SDK to act as a translator between the methods in `providers.Interface` and the simplified methods that provider developers write.
+
+> [!NOTE]  
+> This protocol specification is based on my original [Seasonings plugin protocol RFC](https://github.com/opentofu/opentofu/pull/3051), adapted for the broader OpenTofu provider ecosystem.
+
+## Protocol Overview
+
+The OpenTofu Provider Protocol uses MessagePack-RPC over standard input/output (stdio) streams to enable communication between OpenTofu Core (via the Provider Client Library) and provider implementations. This approach provides:
+
+- **Language agnostic**: Any language with MessagePack support can implement providers
+- **Type preservation**: Maintains OpenTofu's complex type system including unknown values
+- **Efficient serialization**: Binary format with smaller message sizes than JSON
+- **Extensible design**: Protocol can evolve with new message types and capabilities
+
+## Wire Protocol
+
+### RPC Framework Decision
+
+> [!NOTE]  
+> I am currenty not sure if we should propose one of two approaches for the RPC framework:
+> 1. **MessagePack-RPC**: Using the existing [msgpack-rpc specification](https://github.com/msgpack-rpc/msgpack-rpc), though this project appears unmaintained
+> 2. **Custom JSON-RPC-style**: Implementing JSON-RPC semantics but using MessagePack serialization with our own library
+> 
+> Both approaches use MessagePack for serialization to maintain type fidelity and performance, but differ in message structure and ecosystem support.
```

Comment:
Either way seems pretty light compared to the decision of "use messagepack for all encoding and decoding".